controllers/internal/mdutil/util_test.go

```diff
@@ -322,6 +327,131 @@ func TestFindNewMachineSet(t *testing.T) {
+	deploymentRolloutAfterHappened := generateDeployment("nginx")
+	deploymentRolloutAfterHappened.Spec.RolloutAfter = &nextRolloutAfter
+	deploymentRolloutAfterHappened.Annotations[clusterv1.LastRolloutAfterAnnotation] = lastRolloutAfter.Format(time.RFC3339)
+
+	machineSetBeforeLastRolloutAfter := generateMS(deployment)
+	machineSetBeforeLastRolloutAfter.Labels[clusterv1.MachineDeploymentUniqueLabel] = hashValue
+	machineSetBeforeLastRolloutAfter.CreationTimestamp = metav1.Time{
+		Time: lastRolloutAfter.Add(-1 * time.Minute),
+	}
+
+	machineSetAfterLastRolloutAfter := generateMS(deployment)
+	machineSetAfterLastRolloutAfter.Labels[clusterv1.MachineDeploymentUniqueLabel] = hashValue
+	machineSetAfterLastRolloutAfter.CreationTimestamp = metav1.Time{
+		Time: lastRolloutAfter.Add(2 * time.Minute),
+	}
+
+	oldestMachineSetAfterLastRolloutAfter := generateMS(deployment)
+	oldestMachineSetAfterLastRolloutAfter.Labels[clusterv1.MachineDeploymentUniqueLabel] = hashValue
+	oldestMachineSetAfterLastRolloutAfter.CreationTimestamp = metav1.Time{
+		Time: lastRolloutAfter.Add(1 * time.Minute),
+	}
+
+	tests := []struct {
+		Name       string
+		deployment clusterv1.MachineDeployment
+		msList     []*clusterv1.MachineSet
+		expected   *clusterv1.MachineSet
+	}{
+		{
+			Name:       "When the templates match it should find the MachineSet even though the machine-template-hash value differs",
+			deployment: deployment,
+			msList:     []*clusterv1.MachineSet{&newMS, &oldMS},
+			expected:   &newMS,
+		},
+		{
+			Name:       "When there are more than one MachineSet with the same template it should find the oldest",
+			deployment: deployment,
+			msList:     []*clusterv1.MachineSet{&newMS, &oldMS, &newMSDup},
+			expected:   &newMSDup,
+		},
+		{
+			Name:       "When no MachineSets match template it should return nil",
+			deployment: deployment,
+			msList:     []*clusterv1.MachineSet{&oldMS},
+			expected:   nil,
+		},
+		{
+			Name:       "When rolloutAfter is required and no MachineSets were created after it should return nil",
+			deployment: deploymentNeedRolloutAfter,
+			msList:     []*clusterv1.MachineSet{&machineSetBeforeRolloutAfter},
+			expected:   nil,
+		},
+		{
+			Name:       "When rolloutAfter is required it should return the oldest created after rolloutAfter",
+			deployment: deploymentNeedRolloutAfter,
+			msList:     []*clusterv1.MachineSet{&machineSetBeforeRolloutAfter, &machineSetAfterRolloutAfter, &oldestMachineSetAfterRolloutAfter},
+			expected:   &oldestMachineSetAfterRolloutAfter,
+		},
+		{
+			Name:       "When rolloutAfter is not required but another one took place in the past it should return oldest created after the last rolloutAfter",
+			deployment: deploymentRolloutAfterHappened,
+			msList:     []*clusterv1.MachineSet{&machineSetBeforeLastRolloutAfter, &machineSetAfterLastRolloutAfter, &oldestMachineSetAfterLastRolloutAfter},
+			expected:   &oldestMachineSetAfterLastRolloutAfter,
+		},
```

Comment:
I'm a bit confused when would this unit test case would happen. If I understood the code correctly, this use case adds the requirement for the annotation. Wouldn't the `machineSetBeforeLastRolloutAfter` be gone though after the rollout takes place?

Reply:
It won't be necessarily gone, it might be scaled down to zero right. That depends only on .spec.RevisionHistoryLimit right?